tao_compiler/mlir/xla/ral/ral_metadata.cc

```diff
@@ -0,0 +1,162 @@
+
+namespace tao {
+namespace ral {
+
+namespace {
+
+template <typename T>
+bool readFromBinaryStream(std::ifstream& fin, T& t) {
+  if (!fin) return false;
+  fin.read(reinterpret_cast<char*>(&t), sizeof(T));
+  return static_cast<bool>(fin);
+}
+
+template <>
+bool readFromBinaryStream(std::ifstream& fin, std::string& val) {
+  if (!fin) return false;
+  size_t bytes;
+  if (!readFromBinaryStream(fin, bytes)) return false;
+  std::string str(bytes, 0);
+  fin.read(&str[0], bytes);
+  if (!fin) return false;
+  val = str;
+  return true;
+}
+
+template <typename T>
+bool writeToBinaryStream(std::ofstream& out, const T& t) {
+  if (!out) return false;
+  out.write(reinterpret_cast<const char*>(&t), sizeof(T));
+  return static_cast<bool>(out);
+}
+
+template <>
+bool writeToBinaryStream(std::ofstream& out, const std::string& t) {
+  if (!out) return false;
+  if (!writeToBinaryStream(out, t.size())) return false;
+  out.write(t.data(), t.size());
+  return static_cast<bool>(out);
+}
+
+}  // namespace
+
+/* static */ std::unique_ptr<MetadataFile> MetadataFile::loadFromFile(
+    const std::string& filename) {
+  std::unique_ptr<MetadataFile> file(new MetadataFile);
+  std::ifstream fin(filename, std::ios::in | std::ios::binary);
+  fin.seekg(-static_cast<int64_t>(sizeof(size_t)), std::ios::end);
+  size_t numConsts;
+  if (!readFromBinaryStream(fin, numConsts)) return nullptr;
+  fin.seekg(0);
+  for (size_t i = 0; i < numConsts; ++i) {
+    bool isHost;
+    std::string key, value;
+    if (!readFromBinaryStream(fin, isHost)) return nullptr;
+    if (!readFromBinaryStream(fin, key)) return nullptr;
+    if (!readFromBinaryStream(fin, value)) return nullptr;
+    if (isHost) {
+      file->hostConstMap_.emplace(std::move(key), std::move(value));
+    } else {
+      file->deviceConstMap_.emplace(std::move(key), std::move(value));
+    }
+  }
+  return file;
+}
+
+bool MetadataFile::getHostConstant(const std::string& name, std::string& data) {
+  auto it = hostConstMap_.find(name);
+  if (it == hostConstMap_.end()) return false;
+  data = it->second;
+  return true;
+}
+
+bool MetadataFile::releaseHostConstant(const std::string& name) {
+  auto it = hostConstMap_.find(name);
+  if (it == hostConstMap_.end()) return false;
+  hostConstMap_.erase(it);
+  return true;
+}
+
+bool MetadataFile::getDeviceConstant(const std::string& name,
+                                     std::string& data) {
+  auto it = deviceConstMap_.find(name);
+  if (it == deviceConstMap_.end()) return false;
+  data = it->second;
+  return true;
+}
+
+bool MetadataFile::releaseDeviceConstant(const std::string& name) {
+  auto it = deviceConstMap_.find(name);
+  if (it == deviceConstMap_.end()) return false;
+  deviceConstMap_.erase(it);
+  return true;
+}
+
+MetadataFileEmitter::MetadataFileEmitter(const std::string& filename)
+    : filename_(filename) {}
+
+bool MetadataFileEmitter::emitHeader() {
+  out_.open(filename_, std::ios::out | std::ios::binary);
```

Comment:
It helps to to write some magic at the beginning and between parts to do checking.

Reply:
add a magic number. the name of each const includes the md5sum, thus I think it's ok to skip adding checksum again.